src/ethereum/dao_fork/spec.py

```diff
@@ -0,0 +1,779 @@
+"""
+Ethereum Specification
+^^^^^^^^^^^^^^^^^^^^^^
+
+.. contents:: Table of Contents
+    :backlinks: none
+    :local:
+
+Introduction
+------------
+
+Entry point for the Ethereum specification.
+"""
+
+from dataclasses import dataclass
+from typing import List, Optional, Set, Tuple
+
+from ethereum.base_types import Bytes0
+from ethereum.crypto import SECP256K1N
+from ethereum.dao_fork.eth_types import TX_CREATE_COST
+from ethereum.ethash import dataset_size, generate_cache, hashimoto_light
+from ethereum.utils.ensure import ensure
+
+from .. import crypto, rlp
+from ..base_types import U256, U256_CEIL_VALUE, Bytes, Uint
+from . import MAINNET_FORK_BLOCK, vm
+from .bloom import logs_bloom
+from .dao import do_dao_fork
+from .eth_types import (
+    TX_BASE_COST,
+    TX_DATA_COST_PER_NON_ZERO,
+    TX_DATA_COST_PER_ZERO,
+    Address,
+    Block,
+    Bloom,
+    Hash32,
+    Header,
+    Log,
+    Receipt,
+    Root,
+    Transaction,
+)
+from .state import (
+    State,
+    create_ether,
+    destroy_account,
+    get_account,
+    increment_nonce,
+    set_account_balance,
+    state_root,
+)
+from .trie import Trie, root, trie_set
+from .utils.message import prepare_message
+from .vm.interpreter import process_message_call
+
+BLOCK_REWARD = U256(5 * 10 ** 18)
+GAS_LIMIT_ADJUSTMENT_FACTOR = 1024
+GAS_LIMIT_MINIMUM = 5000
+GENESIS_DIFFICULTY = Uint(131072)
+MAX_UNCLE_DEPTH = 6
+
+
+@dataclass
+class BlockChain:
+    """
+    History and current state of the block chain.
+    """
+
+    blocks: List[Block]
+    state: State
+
+
+def apply_fork(old: BlockChain) -> BlockChain:
+    """
+    Transforms the state from the previous hard fork (`old`) into the block
+    chain object for this hard fork and returns it.
+
+    Parameters
+    ----------
+    old :
+        Previous block chain object.
+
+    Returns
+    -------
+    new : `BlockChain`
+        Upgraded block chain object for this hard fork.
+    """
+    do_dao_fork(old.state)
```

Comment:
Any reason to not directly include the implementation here, instead of creating a new function?

Reply:
This way you can work out exactly what the Dao Fork did just by reading the `dao.py` file, rather than spreading the information between `dao.py` and `spec.py`.

In the alternative, putting large constants in `spec.py` seemed inappropriate.

Reply:
I agree that putting the constants in `spec.py` is a bit much, but I think the fork application code should live in `spec.py`. That way `dao.py` just has two constants in it.